File: 2021_fall/알고리즘/12주차/실습/sort.py
from collections import defaultdict, deque

def solve(nodes, graph, pre):
    ans = []
    queue = deque()
    for node in nodes: #모든 노드에 대하여
        if pre[node]==0: queue.append(node) #선행조건이 없으면 큐에 push
    while queue: #큐가 빌 때 까지
        temp = queue.popleft() #큐에서 pop
        ans.append(temp) #정답 배열에 append
        for node in graph[temp]: #모든 연결된 노드에 대하여
            pre[node] -= 1 #연결된 노드의 선행조건 개수를 1 감소
            if pre[node]<=0: queue.append(node) #선행조건이 없다면 큐에 push
    return ans #정답을 리턴

def main():
    n, m = map(int, input().split())
    nodes = input().split()
    graph, pre = defaultdict(list), defaultdict(int) 
    for i in range(m):
        a, b = input().split()
        graph[a].append(b) #a에서 b로 가는 간선을 저장
        pre[b] += 1 #b의 선행조건 개수를 1 증가
    print(*solve(nodes, graph, pre))

if __name__=="__main__": main()

File: 2021_fall/알고리즘/12주차/실습/sort.c
#include <stdio.h>
#include <stdlib.h>

void solve(char ans[], char graph[][26], char pre[], char check[]) {
	char* ansPtr = ans;
	char* queue = malloc(200);
	char* front = queue;
	char* rear = front;
	for(int i=0; i<26; i++) {
		if(pre[i]==0 && check[i]) {
			*rear = i;
			rear++;
		}
	}
	while(front!=rear) {
		char temp = *front;
		front++;
		*ansPtr = (char)temp+65;
		ansPtr++;
		for(int i=0; i<26; i++) {
			if(graph[temp][i]) {
				pre[i]--;
				if(pre[i]<=0) {
					*rear = i;
					rear++;
				}
			}
		}
	}
}

int main() {
	int n, m;
	char a, b;
	scanf("%d %d", &n, &m);
	char graph[26][26] = {0, };
	char pre[26] = {0, };
	char check[26] = {0, };
	for(int i=0; i<m; i++) {
		scanf("%c %c", &a, &b);
		graph[a-65][b-65] = 1;
		pre[b-65]++;
		check[a-65] = 1;
		check[b-65] = 1;
		getchar();
	}
	char* ans = malloc(n);
	solve(ans, graph, pre, check);
	for(int i=0; i<n; i++) {
		printf("%c ", *ans);
		ans++;
	}
	return 0;
}
	

